Create a collection for session, add TTL

Lighty0410 opened the pull request.

Lighty0410 committed:
Add new TTL collection.

Commit:
Create a collection for session, add TTL.

Lighty0410 committed:
Update user.go

Commit:
Transfer createindex to mongo package

Commit:
Merge remote-tracking branch 'origin/ttl-index-collection' into ttl-index-collection

sashayakovtseva commented on pkg/mongo/service.go:
```diff
@@ -23,7 +24,27 @@ func NewService(connectionURL string) (*Service, error) {
+	}
+	err = s.CreateIndex()
+	if err != nil{
+		return s, err
```
retutn nil, fmt.Errorf

Commit:
Change an error handling of CreateIndex

sashayakovtseva commented on pkg/mongo/service.go:
```diff
@@ -23,7 +24,27 @@ func NewService(connectionURL string) (*Service, error) {
+	}
+	err = s.CreateIndex()
+	if err != nil{
+		return s, fmt.Errorf("cannot create an index")
```
and also `...: %v", err)`

Commit:
smt

Commit:
Add an expiration time for mongo files

Commit:
Add an expiration date to mongo, delete it from cookie

Commit:
Add a special struct to store a session in mongoDB

Lighty0410 committed:
Update main.go

sashayakovtseva commented:
Did you run `make lint && make test`?

sashayakovtseva commented on pkg/mongo/service.go:
```diff
@@ -23,7 +25,30 @@ func NewService(connectionURL string) (*Service, error) {
+func (s *Service) CreateIndex() error {
+	var opt options.IndexOptions
+	opt.SetExpireAfterSeconds(60*5)
+	opt.SetUnique(true)
```
Why unique?

sashayakovtseva commented on pkg/mongo/service.go:
```diff
@@ -23,7 +25,30 @@ func NewService(connectionURL string) (*Service, error) {
+	var opt options.IndexOptions
+	opt.SetExpireAfterSeconds(60*5)
+	opt.SetUnique(true)
+	keys := bsonx.Doc{{Key: "expiration", Value: bsonx.Int32(int32(1))}}
```
Key is 'expiration', but TTL is set to 5 mins, which means object will be removed at expiration+5m. A bit confusing name for key then.

sashayakovtseva replied:
can be simplified to `keys := bson.M{"expiration": 1}`

sashayakovtseva commented on pkg/mongo/service.go:
```diff
@@ -3,9 +3,11 @@ package mongo
 import (
 	"context"
 	"fmt"
+	"github.com/mongodb/mongo-go-driver/x/bsonx"
```
you didn't run `make lint`

sashayakovtseva replied:
also I think you can use `github.com/mongodb/mongo-go-driver/bson` here

sashayakovtseva commented on pkg/mongo/user.go:
```diff
@@ -15,6 +15,12 @@ type User struct {
 
+type Session struct {
+	Name string `bson:"name"`
+	CookieHash string `bson:"cookiehash"`
```
I'd like mongo not to know anything about cookie, even in struct names.

sashayakovtseva commented on pkg/mongo/user.go:
```diff
@@ -30,6 +36,15 @@ func (s *Service) AddUser(u *User) error {
+	c := s.db.Collection("session")
+	_, err := c.InsertOne(context.Background(), u)
+	if err != nil {
+		return fmt.Errorf("cannot create a session")
```
wrap error!

sashayakovtseva commented on pkg/server/crypto.go:
```diff
@@ -16,3 +19,10 @@ func generateHash(password string) (string, error) {
+		return ""
+	}
+	return base64.URLEncoding.EncodeToString(b)
+}
```
no line feed

sashayakovtseva commented on pkg/server/crypto.go:
```diff
@@ -16,3 +19,10 @@ func generateHash(password string) (string, error) {
 func compareHash(hash string, password []byte) error {
 	return bcrypt.CompareHashAndPassword([]byte(hash), password)
 }
+func generateCookieHash() string {
```
generateToken is better

sashayakovtseva commented on pkg/server/register.go:
```diff
@@ -62,5 +62,20 @@ func (s *EkadashiServer) handleLogin(w http.ResponseWriter, r *http.Request) {
 	}
+	cookieHash := generateCookieHash()
+	expires := time.Now()
+	cookie := http.Cookie{
```
In general ok here, but as a style preference I think you should create session object first, insert it into mongo and then form cookie based on session object

sashayakovtseva commented on pkg/server/register.go:
```diff
@@ -62,5 +62,20 @@ func (s *EkadashiServer) handleLogin(w http.ResponseWriter, r *http.Request) {
+	cookieHash := generateCookieHash()
+	expires := time.Now()
+	cookie := http.Cookie{
+		Name:req.Username,
```
name of a cookie, not of a user, should be smth like 'session_id'

sashayakovtseva commented on pkg/server/register.go:
```diff
@@ -62,5 +62,20 @@ func (s *EkadashiServer) handleLogin(w http.ResponseWriter, r *http.Request) {
+		Expiration:expires,
+	})
+	if err != nil{
+		jsonError(w, http.StatusInternalServerError, fmt.Errorf("cannot create a cookie"))
```
wrap error!!!

Commit:
Wrap an error, changed the field in Session structure

Commit:
Refactor the minor piece of code for better look

1. Add wrapers for error both in register and user
2. Add a special structure to set session
3. Refactor CreateIndex for better look

Commit:
Merge remote-tracking branch 'origin/ttl-index-collection' into ttl-index-collection

Commit:
Set cookie settings from mongo.session instead of structure inside server

Commit:
Delete session struct from server

sashayakovtseva commented on pkg/mongo/service.go:
```diff
@@ -23,7 +25,29 @@ func NewService(connectionURL string) (*Service, error) {
+	c := s.db.Collection("session")
+	_, err := c.Indexes().CreateOne(context.Background(), model)
+	if err != nil {
+		return fmt.Errorf("cannont create session: %v",err)
```
cannot create session index

sashayakovtseva commented on pkg/server/register.go:
```diff
@@ -59,8 +60,22 @@ func (s *EkadashiServer) handleLogin(w http.ResponseWriter, r *http.Request) {
+		jsonError(w, http.StatusInternalServerError, fmt.Errorf("cannot create a cookie: %v",err))
+	}
+	cookieSettings := &mongo.Session{}
+	cookie := http.Cookie{
```
You are setting empty cookie here! should be
```
userSession:=&mongo.Session{
		Name:req.Username,
		SessionHash:generateToken(),
		LastModifiedDate:time.Now(),
	}
err = s.db.CreateSession(userSession)
	if err != nil{
		jsonError(w, http.StatusInternalServerError, fmt.Errorf("cannot create session: %v",err))
	}
	cookie := http.Cookie{
		Name: "session_token",
		Value: userSession.SessionHash,
	}
	http.SetCookie(w,&cookie)```

sashayakovtseva commented on pkg/server/register.go:
```diff
@@ -59,8 +60,22 @@ func (s *EkadashiServer) handleLogin(w http.ResponseWriter, r *http.Request) {
+		LastModifiedDate:time.Now(),
+	})
+	if err != nil{
+		jsonError(w, http.StatusInternalServerError, fmt.Errorf("cannot create a cookie: %v",err))
```
cannot create session, not cooke

Commit:
Fix an error message in CreateIndex function

Commit:
Set a  proper structure for CreateSession, set a cookie based on it

1. And also i had watched what's a cookie name on this purpose

sashayakovtseva commented:
All ok, awaiting linter only

Lighty0410 committed:
Refactor code by metalinter
